### PR TITLE
#1264 Port Shapes Lab to WPF

### LINKED ISSUE #1264: Make the shapes lab window take full vertical space
Received from a user:
For the Shapes Lab, would it be possible to have it use the fully vertical height of the window, like the Resize Lab does? I have many shapes in a category, but it's only using about half the screen.

Additionally, could we specify the thumbnail size and text size? Since I have a lot of shapes, I'd like to squeeze as many in as possible.

### PR DESCRIPTION
**Pull request by YuPeiHenry**
Partially fixes #1264. Next PR will add options to set shape icon size/ text size in Shapes Lab.

Ready for review. Tested and Working on PPT2016.

**Outline of Solution**
* List of shapes uses ListBox and WrapPanel
* Context Menus use native WPF functionality
* Shape storage code was adapted from old shapes lab.
* Multi-selection is natively handled by WPF's WrapPanel.
* New code is shorter.

Features tested (with multiple Powerpoints open):
* Adding shapes to ShapesLab
* Deleting shapes from ShapesLab
* Renaming shapes in ShapesLab
* Moving shapes in ShapesLab
* Adding shapes to Slide from ShapesLab
* Adding/removing categories
* Setting category as default
* Importing shapes
* Changing shapes lab settings (Note that migration has some pre-existing issues).

Some differences from original shapes lab pane:
* No left-click rename functionality (like renaming files in Windows).
* No text in the shape pane when there are no shapes in the category.

### CODE REVIEW

**Comment by YuPeiHenry** [comment]
@blewjy @ChesterSng 
The following have been fixed:
* `Move Shapes To` now shows the current category names and lives in the context menu.
* Whitespace has been filled with link to download shapes (not sure where did the link disappear to in the non-WPF)
* Visual cue added for default folder
* Settings moved to last option in context menu.
* User input is trimmed for both shape name and category name.
* Renaming a category no longer results in a bug.
* Icons for shapes lab are no longer missing.
* Shape textbox removed (main issue was that there is no convenient way to focus on an entire WPF element from the PowerPoint. Removing the textbox is kind of altering the old implementation though).
* Tooltip is removed.

**Comment by ChesterSng** [comment]
Hi @YuPeiHenry, thanks for making the changes. Some bugs I found: 

**1.) Move shape to can only move shape to "My Shapes" category**
i. Create at least 1 other category 
ii. Select shape in "My Shapes" and right-click
iii. Move Shape To --> Only My Shapes will appear

Furthermore, 
**2.) After "My Shapes" category is deleted, `Move Shape To` now has all categories**
i. From step 1, create one more category (for better illustration)
ii. Delete "My Shapes" category
iii. Now go to any category and use the same `Move Shape To`
iv. The options will now show new category 1, new category 2 and "My Shapes" 
v. Selecting "My Shapes" will throw an exception

**Also, there seems to be a duplicate `Move Shapes To` option.**

**3.) Renaming of category**
i. Try to rename a category to the same name as another category 
ii. This behaviour is rejected (Expected) 
iii. Use the same name as another category and add a space behind, i.e. "My Shapes "
iv. Unhandled exception is thrown

**Comment by YuPeiHenry** [comment]
Oh I forgot to remove the older implementation.

**Comment by YuPeiHenry** [comment]
@ChesterSng I have fixed the 3 issues you have found. Thanks!

File: PowerPointLabs/Test/FunctionalTest/ShapesLabTest.cs [hunk 1]
```diff
@@ -138,27 +135,25 @@ private void TestSavePlaceholderToShapesLabWithAddShapesButton(IShapesLabControl
             PpOperations.SelectSlide(shapesSlideNum);
             ExpectAddShapeButtonDisabled(shapesLab);
             PpOperations.SelectShapesByPrefix("Placeholder");
-            ExpectAddShapeButtonEnabled(shapesLab);
+            ExpectAddShapeButtonDisabled(shapesLab);
```
Comment (leeyh20) [inline PowerPointLabs/Test/FunctionalTest/ShapesLabTest.cs]: is the function `ExpectAddShapeButtonEnabled(shapesLab)` unused in this script?

Comment (leeyh20) [reply PowerPointLabs/Test/FunctionalTest/ShapesLabTest.cs]: line 171

Comment (YuPeiHenry) [reply PowerPointLabs/Test/FunctionalTest/ShapesLabTest.cs]: @leeyh20 I believe it is used. But selecting the (empty) placeholder does not seem to enable the button.

File: PowerPointLabs/PowerPointLabs/ShapesLab/Views/ShapesLabCategoryInfoDialogBox.xaml.cs [hunk 2]
```diff
@@ -27,25 +27,27 @@ public partial class ShapesLabCategoryInfoDialogBox
 
         // Regex = [<>:"/\\|?*]
         private const string InvalidCharsRegex = "[<>:\"/\\\\|?*]";
+        private bool shouldUseExistingCategory;
```
[the remark this answers is not in the document]
Comment (YuPeiHenry) [reply PowerPointLabs/PowerPointLabs/ShapesLab/Views/ShapesLabCategoryInfoDialogBox.xaml.cs]: @leeyh20 This is no longer used. Previously when I ported shapes lab over, moving a shape to a different category also used this dialogbox, hence I needed an option to choose whether the category should be an existing one.

[the remark this answers is not in the document]
Comment (leeyh20) [reply PowerPointLabs/PowerPointLabs/ShapesLab/Views/ShapesLabCategoryInfoDialogBox.xaml.cs]: Oh i see.

**Comment by leeyh20** [comment]
Need to resolve conflicts.

**Comment by YuPeiHenry** [comment]
@leeyh20 I have resolved the conflict.

File: PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/AddCustomShape/AddCustomShapeImageHandler.cs [hunk 3]
```diff
@@ -4,10 +4,10 @@
 using PowerPointLabs.ActionFramework.Common.Interface;
 using PowerPointLabs.TextCollection;
 
-namespace PowerPointLabs.ActionFramework.ShapesLab
+namespace PowerPointLabs.ActionFramework.ShortcutsLab
```
Comment (leeyh20) [inline PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/AddCustomShape/AddCustomShapeImageHandler.cs]: Sorry just spotted this.
Is the namespace supposed to be ShortcutsLab or ShapesLab?

Comment (leeyh20) [reply PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/AddCustomShape/AddCustomShapeImageHandler.cs]: Ohh... did it change to ShortcutsLab?

Comment (YuPeiHenry) [reply PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/AddCustomShape/AddCustomShapeImageHandler.cs]: It seems the other shortcuts are using `ShortcutsLab` namespace. I'm not sure which would be more appropriate.

Comment (blewjy) [reply PowerPointLabs/PowerPointLabs/ActionFramework/ShortcutsLab/AddCustomShape/AddCustomShapeImageHandler.cs]: I believe `ShortcutsLab` will be more appropriate here given that it's a feature under ShortcutsLab

**Comment by leeyh20** [comment]
Thanks for this PR! @YuPeiHenry

**Pull request merged**